firedrake/preconditioners/hiptmair.py

```diff
@@ -0,0 +1,262 @@
+                               save=False):
+            coarse_solver.setFromOptions()
+
+    def update(self, pc):
+        self._assemble_coarse_op()
+        self.pc.setUp()
+
+    def apply(self, pc, X, Y):
+        dm = self._dm
+        with dmhooks.add_hooks(dm, self, appctx=self._ctx_ref):
+            self.pc.apply(X, Y)
+
+    def applyTranspose(self, pc, X, Y):
+        dm = self._dm
+        with dmhooks.add_hooks(dm, self, appctx=self._ctx_ref):
+            self.pc.applyTranspose(X, Y)
+
+    def view(self, pc, viewer=None):
+        super(TwoLevelPC, self).view(pc, viewer)
+        if hasattr(self, "pc"):
+            viewer.printfASCII("Two level PC\n")
+            self.pc.view(viewer)
+
+
+class HiptmairPC(TwoLevelPC):
+
+    _prefix = "hiptmair_"
+
+    def coarsen(self, pc):
+        from firedrake_citations import Citations
+        from firedrake import FunctionSpace, TestFunction, TrialFunction
+        from firedrake.interpolation import Interpolator
+        from ufl.algorithms.ad import expand_derivatives
+        from ufl import replace, zero, grad, curl, as_vector
+
+        Citations().register("Hiptmair1998")
+        appctx = self.get_appctx(pc)
+        V = dmhooks.get_function_space(pc.getDM())
+
+        _, P = pc.getOperators()
+        if P.getType() == "python":
+            ctx = P.getPythonContext()
+            a = ctx.a
+            bcs = tuple(ctx.bcs)
+        else:
+            ctx = dmhooks.get_appctx(pc.getDM())
+            problem = ctx._problem
+            a = problem.Jp or problem.J
+            bcs = tuple(problem.bcs)
+
+        mesh = V.mesh()
+        element = V.ufl_element()
+        degree = element.degree()
+        try:
+            degree = max(degree)
+        except TypeError:
+            pass
+        formdegree = V.finat_element.formdegree
+        if formdegree == 1:
+            celement = curl_to_grad(element)
+            dminus = grad
+            G_callback = appctx.get("get_gradient", None)
+        elif formdegree == 2:
+            celement = div_to_curl(element)
+            dminus = curl
+            if V.shape:
+                dminus = lambda u: as_vector([curl(u[k, ...]) for k in range(u.ufl_shape[0])])
+            G_callback = appctx.get("get_curl", None)
+        else:
+            raise ValueError("Hiptmair decomposition not available for", element)
+
+        coarse_space = FunctionSpace(mesh, celement)
+        assert coarse_space.finat_element.formdegree + 1 == formdegree
+        coarse_space_bcs = tuple([bc.reconstruct(V=coarse_space, g=0) for bc in bcs])
+
+        # Get only the zero-th order term of the form
+        beta = replace(expand_derivatives(a), {grad(t): zero(grad(t).ufl_shape) for t in a.arguments()})
+
+        test = TestFunction(coarse_space)
+        trial = TrialFunction(coarse_space)
+        coarse_operator = beta(dminus(test), dminus(trial), coefficients={})
+
+        if formdegree > 1 and degree > 1:
```

Comment:
If someone sets a shift, I think it should be used, even if degree == 1 (we set mechanism, not policy)

Reply:
It could be that the user might want to set the shift only for the high order space in a p-multgrid hierarchy. Should the shift be a function of the degree (in a similar way the callback to tabulate the gradient is a function of the spaces and bcs)?

Reply:
Ah, hmm, I see. In that case I guess it makes sense to just code the sensible thing, as you have done. Making the user supply a shift function that depends on degree seems overly complicated.